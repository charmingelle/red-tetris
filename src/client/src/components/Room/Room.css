.room {
  width: 100vw;
  height: 100vh;
  display: flex;
}

.field-container {
  width: 70%;
  height: 100vh;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  background-color: #666666;
}

.field {
  position: relative;
  width: 500px;
  height: 1000px;
  outline: 5px solid #f9f9f9;
  background-color: black;
}

.my-score {
  margin-top: 30px;
  font-size: 36px;
  color: #f9f9f9;
}

.game-details-container {
  width: 30%;
  height: 100vh;
  overflow: scroll;
  background-color: #333;
}

.game-details {
  padding: 5px;
  display: flex;
  flex-direction: column;
  align-items: center;
}

.start-game-button {
  background-color: #9901cc;
  border: 5px solid;
  border-left-color: #b804e6;
  border-top-color: #b804e6;
  border-right-color: #660099;
  border-bottom-color: #660099;
  color: #fff;
  font-size: 30px;
  font-weight: bold;
  padding: 10px;
}

.game-over {
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  background-color: #c00;
  color: #fff;
  font-size: 36px;
  font-weight: bold;
  padding: 15px;
  border: 5px solid white;
}
