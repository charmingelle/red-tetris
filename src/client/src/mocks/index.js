export const RED = {
  main: '#c00',
  lighter: '#f00',
  darker: '#980000',
};

export const LINE = [
  [0, 0, 1, 0],
  [0, 0, 1, 0],
  [0, 0, 1, 0],
  [0, 0, 1, 0],
];

export const GI = [
  [0, 0, 0],
  [1, 1, 1],
  [0, 0, 1],
];

export const EMPTY_PILE = [
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
];

export const NOT_EMPTY_PILE = [
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [RED, RED, RED, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, RED, 0, 0, 0, 0, 0, 0, 0],
];

export const GAME_OVER_PILE = [
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, RED, RED, RED, RED, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [RED, RED, RED, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, RED, 0, 0, 0, 0, 0, 0, 0],
];

export const PILE_WITH_ROWS_TO_REMOVE = [
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [RED, RED, RED, RED, RED, RED, RED, RED, RED, RED],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [RED, RED, RED, RED, RED, RED, RED, RED, RED, RED],
  [RED, RED, RED, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, RED, 0, 0, 0, 0, 0, 0, 0],
];

export const TETRO = {
  figure: [GI],
  row: 5,
  col: 4,
  color: RED,
};

export const ROOM = {
  id: 'Room',
  leader: 'id-0',
  players: {
    'id-0': {
      id: 'id-0',
      name: 'Anna',
      pile: EMPTY_PILE,
      tetro: null,
      score: 0,
      isGameOver: false,
    },
  },
};

export const EMPTY_STORE = {
  socket: null,
  myId: null,
  myName: null,
  myRoom: null,
  people: {},
  rooms: {},
  tetro: null,
};

export const PERSON_ANNA = { name: 'Anna', score: 0 };

export const PERSON_OLYA = { name: 'Olya', score: 0 };
