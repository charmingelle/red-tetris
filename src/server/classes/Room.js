const Player = require('./Player');
const Game = require('./Game');

module.exports = class Room {
  constructor({ id, leader, players = {} }) {
    this.id = id;
    this.leader = leader;
    this.players = players;
    this.game = null;
  }

  addPlayer(playerId, playerName) {
    this.players[playerId] = new Player({
      id: playerId,
      name: playerName,
    });
  }

  removePlayer(playerId) {
    delete this.players[playerId];
  }

  setLeader(leader) {
    this.leader = leader;
  }

  startGame() {
    this.game = new Game(this);
  }

  endGame() {
    this.game = null;
  }

  getRoomData() {
    const clientPlayers = {};

    Object.keys(this.players).map(
      playerId =>
        (clientPlayers[playerId] = this.players[playerId].getPlayerData()),
    );
    return {
      id: this.id,
      leader: this.leader,
      players: clientPlayers,
      game: this.game ? this.game.getGameData() : null,
    };
  }
};
