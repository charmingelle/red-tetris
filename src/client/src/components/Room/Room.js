import React from 'react';
import { connect } from 'react-redux';
import './Room.css';
import { Tetromino } from '../Tetromino';
import {
  moveTetro,
  moveTetroDown,
  dropTetro,
  rotateTetro,
} from '../../actions';
import { LEFT, RIGHT, DOWN, UP } from '../../constants';
import { Pile } from '../Pile';
import { store } from '../../index';
import { Others } from '../Others';
import { Penalty } from '../Penalty';

const keyDownHandler = ({ key }) => {
  switch (key) {
    case LEFT: {
      store.dispatch(moveTetro({ left: -1, top: 0 }));
      break;
    }
    case RIGHT: {
      store.dispatch(moveTetro({ left: 1, top: 0 }));
      break;
    }
    case UP: {
      store.dispatch(rotateTetro());
      break;
    }
    case DOWN: {
      store.dispatch(dropTetro());
      break;
    }
    default: {
      return null;
    }
  }
};

window.setInterval(() => store.dispatch(moveTetroDown()), 750);

window.addEventListener('keydown', keyDownHandler);

const startGame = (socket, roomId) => () => {
  socket.emit('start-game', {
    roomId,
  });
};

const StartGameButton = ({ socket, myId, roomId, leader, game }) =>
  myId === leader && !game ? (
    <button className="start-game-button" onClick={startGame(socket, roomId)}>
      START
    </button>
  ) : null;

const GameDetails = ({ socket, myId, roomId, leader, game }) =>
  roomId ? (
    <div className="game-details">
      <StartGameButton
        socket={socket}
        myId={myId}
        roomId={roomId}
        leader={leader}
        game={game}
      />
      <Others />
    </div>
  ) : null;

const Game = ({ game }) =>
  game ? (
    <>
      <Tetromino />
      <Pile />
      <Penalty />
      {game.isOver && <div className="game-over">GAME OVER</div>}
    </>
  ) : null;

const Field = ({ myName, game, score }) => (
  <div className="field-container">
    <div className="field" tabIndex={0}>
      <Game game={game} />
    </div>
    <div className="my-score">{`${myName}: ${score}`}</div>
  </div>
);

const RoomInner = ({ socket, myId, roomId, myName, leader, game, score }) => (
  <div className="room">
    <Field myName={myName} game={game} score={score} />
    <div className="game-details-container">
      <GameDetails
        socket={socket}
        myId={myId}
        roomId={roomId}
        leader={leader}
        game={game}
      />
    </div>
  </div>
);

const mapStateToProps = ({ socket, myId, myRoomId, myName, rooms }) => ({
  socket,
  myId,
  roomId: myRoomId,
  myName,
  leader: rooms[myRoomId].leader,
  game: rooms[myRoomId].game,
  score: rooms[myRoomId].players[myId].score,
});

export const Room = connect(mapStateToProps)(RoomInner);
