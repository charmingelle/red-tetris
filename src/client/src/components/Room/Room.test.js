import React from 'react';
import { Room } from './Room';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { ROOM } from '../../mocks';

const mockStore = configureStore([]);

test('Room', () => {
  const store = mockStore({
    socket: null,
    myId: 'id-0',
    myName: 'Anna',
    myRoom: ROOM,
    people: {},
    rooms: {},
    tetro: null,
  });

  expect(
    renderer
      .create(
        <Provider store={store}>
          <Room />
        </Provider>,
      )
      .toJSON(),
  ).toMatchSnapshot();
});
