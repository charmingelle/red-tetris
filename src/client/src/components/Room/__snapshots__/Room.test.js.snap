// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Room 1`] = `
<div
  className="room"
>
  <div
    className="field-container"
  >
    <div
      className="field"
      style={
        Object {
          "height": "800px",
          "width": "400px",
        }
      }
      tabIndex={0}
    />
    <div
      className="my-score"
    >
      Anna: 0
    </div>
  </div>
  <div
    className="game-details-container"
  >
    <div
      className="game-details"
    >
      <button
        className="start-game-button"
        onClick={[Function]}
      >
        START
      </button>
      <ul
        className="others"
      />
    </div>
  </div>
</div>
`;
